--- a/dpa_app-4.03.0/dpa.c
+++ b/dpa_app-4.03.0/dpa.c
@@ -768,8 +768,8 @@
    	retval = fmc_compile(&cmodel, cfg_file, pcd_file, pdl_file, sp_file,  SP_OFFSET, 0,
 			NULL);
 	if (retval) {
-		printf("%s::unable to compile fmc input files, err %d\n",
-			__FUNCTION__, retval);
+		printf("%s::unable to compile fmc input files, err %d: %s\n",
+			__FUNCTION__, retval, fmc_get_error());
                 return -1;
 	}
 	retval = -1;
